kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt

```diff
@@ -115,49 +144,60 @@ class AppEngineEventStore(private val kind: String = "Event", private val messag
             val eventsModel = events.mapIndexed { index, it -> EventModel(it.kind, currentVersion + index + 1, it.identityId, it.timestamp, it.data.payload.toString(Charsets.UTF_8)) }
 
             val sequenceIds = (1..eventsModel.size).map { idGenerator.nextId() }
+
             val eventIndexes = eventsModel.mapIndexed { index, eventModel ->
                 val sequenceId = sequenceIds[index]
-                val eventIndex = Entity(indexKind, eventModel.version, aggregateKey)
-                eventIndex.setUnindexedProperty(aggregateIdProperty, aggregateId)
-                eventIndex.setIndexedProperty(aggregateTypeProperty, aggregateType)
-                eventIndex.setUnindexedProperty(aggregateIndexProperty, aggregateIndex)
-                eventIndex.setUnindexedProperty(versionProperty, eventModel.version)
-                eventIndex.setIndexedProperty(sequenceProperty, sequenceId)
-                eventIndex
-            }
 
-            val eventsAsText = eventsModel.map { Text(messageFormat.formatToString(it)) }
+                val indexKey = keyQueryFactory.createIndexKey(aggregateKey, eventModel)
+
+                val eventIndex = Entity.newBuilder(indexKey)
+                eventIndex.set(aggregateIdProperty, StringValue.newBuilder(aggregateId).setExcludeFromIndexes(true).build())
+                eventIndex.set(aggregateTypeProperty, aggregateType)
+                eventIndex.set(aggregateIndexProperty, LongValue.newBuilder(aggregateIndex).setExcludeFromIndexes(true).build())
+                eventIndex.set(versionProperty, LongValue.newBuilder(eventModel.version).setExcludeFromIndexes(true).build())
+                eventIndex.set(sequenceProperty, sequenceId)
+                eventIndex.build()
+            }.toTypedArray()
+
+            var sizeInBytes = 0
+            val eventsAsText = eventsModel.map {
+                val eventPayload = messageFormat.formatToString(it)
+                sizeInBytes += eventPayload.toByteArray().size
```

Comment:
You have sum the size of the events. I'm afraid that the format of the entity and it's size will be a little bit bigger,, because of the structure. 
```
{
  "values": [
    {
      "meaning": 15,
      "stringValue": "{\"kind\":\"AccountCreatedEvent\",\"version\":1,\"identityId\":\"4911190124789760\",\"timestamp\":1559121627682,\"payload\":\"{\\\"accountId\\\":\\\"1yXz11D\\\",\\\"userId\\\":\\\"4911190124789760\\\",\\\"currency\\\":\\\"BGN\\\"}\"}",
      "excludeFromIndexes": true
    },
    {
      "meaning": 15,
      "stringValue": "{\"kind\":\"DepositMadeEvent\",\"version\":2,\"identityId\":\"4911190124789760\",\"timestamp\":1559627341102,\"payload\":\"{\\\"sourceId\\\":\\\"djKoD03KR8Q\\\",\\\"transactionId\\\":\\\"djKoD03KR8Q\\\",\\\"sourceType\\\":\\\"PAYMENT_DEPOSIT\\\",\\\"accountId\\\":\\\"1yXz11D\\\",\\\"amount\\\":35.7,\\\"initialBalance\\\":0.0}\"}",
      "excludeFromIndexes": true
    },
    {
      "meaning": 15,
      "stringValue": "{\"kind\":\"DepositMadeEvent\",\"version\":3,\"identityId\":\"4911190124789760\",\"timestamp\":1559629245334,\"payload\":\"{\\\"sourceId\\\":\\\"LoyK26a32yL\\\",\\\"transactionId\\\":\\\"LoyK26a32yL\\\",\\\"sourceType\\\":\\\"PAYMENT_DEPOSIT\\\",\\\"accountId\\\":\\\"1yXz11D\\\",\\\"amount\\\":105.74,\\\"initialBalance\\\":35.7}\"}",
      "excludeFromIndexes": true
    },
```

Did you test this? I'm afraid that an exception will be thrown and the snapshot will not be create

Reply:
The AppEngineEventstoreTest contains tests for large events and entities and one additional test is added for adding of many events.

The entity check now skips 100 kb for metadata to ensure that the entity will be saved. Yes, there is a case when entity could not be saved but the datastore will throw an error and either events or snapshot will be saved as everything is in a transactional block and client will be notified with an internal error.

Reply:
ok